### PR TITLE
Batching rule for torch.mul

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40097 Add batching rule for torch.expand
* #40042 Change VmapTransforms to use SmallVector instead of `vector<int64_t>`
* #40028 Implement tensor.size(int) for BatchedTensor
* **#39859 Batching rule for torch.mul**

This PR implements the batching rule for `torch.mul` for the (Tensor,
Tensor) overload.

NB: ~250 lines of this PR are tests, so please don't be scared away by
the line count.

It introduces the BroadcastingVmapTransform, which is the VmapTransform
one should use for operations that broadcast their inputs. This
transform:
- permutes all batch dimensions to the front of the tensors
- aligns the batch dimensions of the tensors, adding extra 1's where
necessary
- aligns the non-batch dims of the tensors, adding extra 1's where
necessary.

Test Plan:
- Test BroadcastingVmapTransform in `./build/bin/vmap_test`.
- Test mul_batching_rule in `./build/bin/vmap_test`.

Differential Revision: [D22067337](https://our.internmc.facebook.com/intern/diff/D22067337)

### CODE REVIEW

File: aten/src/ATen/VmapTransforms.cpp [hunk 1]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
   return bdims;
 }
 
-static bool batchDimsAreSameSize(
```
Comment (zou3519) [inline aten/src/ATen/VmapTransforms.cpp:97]: The diff view in github doesn't make it clear what happened, so I am explaining it here.

Previously, `VmapPhysicalView::newLogicalFromPhysical(physical)` would check that leading dimensions of `physical` have the same sizes as the batch dimensions of VmapPhysicalView's tensor.

However, I wanted to reuse `VmapPhysicalView` as the output for `BroadcastingVmapTransform`. BroadcastingVmapTransform returns physical views that don't have to have the same batch dimension sizes because we know that the underlying operation is able to broadcast the dimensions anyways. For example, given inputs of size (B0,) and (B0, B1), where B0 and B1 are batch dimensions, `BroadcastingVmapTransform::logicalToPhysical` returns two tensors of size `(B0, 1)` and `(B0, B1)`. Later, the broadcasting operation (like `torch::mul`) can happily multiply two tensors of those size together because it handles the broadcasting under the hood.

File: aten/src/ATen/VmapTransforms.cpp [hunk 2]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
+  for (const auto& tensor : logical_tensors) {
+    // NB: It's possible that we didn't actually need to align `tensor`.
+    // For example, when adding two tensors of size (B, 2), and (3, 2), where
+    // the first Tensor is a BatchedTensor with batch dim B and the second is
+    // a regular Tensor, we will return views of size (B, 1, 2) and (1, 3, 2).
+    // However, the view on the second tensor is unnecessary: broadcasting
+    // semantics allow for the addition of two tensors of size (B, 1, 2) and (3, 2)!
+    //
+    // If this unnecessary view is a problem, consider optimizing it away in
+    // the future. This may involve creating a new type of VmapPhysicalView
```
Comment (zou3519) [inline aten/src/ATen/VmapTransforms.cpp:198]: This comment here also suggests that perhaps BroadcastingVmapTransform should return a different kind of VmapPhysicalView. Right now, `VmapPhysicalView` must hold a `levels` bitset to perform the physical->logical map, but that `levels` bitset is also used to map logical (dims, shapes, etc) to physical (dims, shapes etc).

An alternative might be to separate the VmapPhysicalView::newLogicalFromPhysical from VmapPhysicalView

File: aten/src/ATen/VmapTransforms.cpp [hunk 3]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
+    if (ndim > largest_example_dim) {
+      largest_example_dim = ndim;
     }
   }
-  return true;
+  return { levels, largest_example_dim };
 }
 
-Tensor VmapPhysicalView::newLogicalFromPhysical(const Tensor& physical) {
+std::vector<VmapPhysicalView> BroadcastingVmapTransform::logicalToPhysical(TensorList logical_tensors) {
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.cpp:177]: Allocation of a vector here is irritating. Maybe a SmallVector here would be better.

Comment (zou3519) [reply aten/src/ATen/VmapTransforms.cpp:177]: If performance matters, then I'm planning on having an overload that takes in two tensors and returns a pair of tensors. The most input arguments a broadcasting operators in pytorch is 3 I think, so we could just have an overload for two tensors and one for three tensors.

I'll change this to smallvector for now though

File: aten/src/ATen/VmapTransforms.cpp [hunk 4]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
   }
-  for (int64_t dim = 0; dim < num_batch_dims; dim++) {
-    if (physical_sizes[dim] != reference_sizes[dim]) {
-      return false;
+
+  return physical_tensor.view(aligned_sizes);
+}
+
+static std::pair<std::bitset<kVmapNumLevels>,int64_t>
+getLevelsAndLargestExampleDim(TensorList logical_tensors) {
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.cpp:160]: What's an "example dim"?

File: aten/src/ATen/VmapTransforms.cpp [hunk 5]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
+  TORCH_INTERNAL_ASSERT(logical_tensors.size() > 0);
+  std::bitset<kVmapNumLevels> levels;
+  int64_t largest_example_dim = -1;
+  for (const auto& tensor : logical_tensors) {
+    auto* batched = maybeGetBatched(tensor);
+    if (batched) {
+      levels = levels | createLevelsBitset(batched->bdims());
+    }
+    auto ndim = tensor.dim();
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.cpp:169]: It is probably best to mention explicitly here that `dim()` here doesn't include the batch dimensions

File: aten/src/ATen/VmapTransforms.cpp [hunk 6]
```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
+    return {permuteBatchDimsToFront(batched), createLevelsBitset(batched->bdims())};
   }
-  if (reference_sizes.size() < num_batch_dims) {
-    return false;
+  return {self, 0};
+}
+
+// Given a Tensor or a BatchedTensor, creates a physical view of the tensor
+// such that it has a batch dimension for each levels in `requested_levels`
+// and `requested_example_dim` number of non-batch-dimensions.
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.cpp:115]: It will be easier to understand if we explain why you want to do this. An example would help too.

File: aten/src/ATen/VmapTransforms.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/VmapTransforms.cpp:160]: I'm going to rename this to "LogicalDim" to align with the current vocabulary as much as possible. But in the other function it doesn't make sense to call it a "logical dim", so I'll add a comment there.

File: aten/src/ATen/VmapTransforms.cpp [hunk 7]
```diff
@@ -94,33 +94,128 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
...
+  std::copy(
+      physical_sizes.rbegin(),
+      physical_sizes.rbegin() + tensor_example_dim,
+      aligned_sizes.rbegin());
+
+  // align the bdims
+  int64_t level = 0;
+  int64_t dim = 0;
+  int64_t tensor_dim = 0;
+  for (int64_t level = 0; level < kVmapNumLevels; level++) {
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.cpp:162]: Something that made this a little hard to understand for me was the fact that you were iterating over *all* 64 levels.  Intuitively to me, I would have expected to process only the union of levels which actually existed (aka, the loop counter ought to be `dim`). This is not a suggestion to change the implementation per se; this may very well be the most convenient way to go about doing it as things are written, but I did have to do a little mental leap here to figure out what was going on.

File: aten/src/ATen/VmapTransforms.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/VmapTransforms.cpp:214]: Hot take: this optimization doesn't seem worth it.

**Review by ezyang — APPROVED** [review]
The code here seems correct.

One thing that I feel like would be quite useful for testing and understandability is a non-optimized reference implementation of the transforms, written recursively. What I am imagining is that you have some sort of representation where you maintain dimensions sorted in level order, with levels directly attached to each dimension (rather than indirectly represented via the bitset as you have here.) Then aligning for broadcasting simply involves recursively processing a pair of tensors and inserting ones whenever there is a level skip. It should be a pretty short program and hard to get wrong.

File: aten/src/ATen/VmapTransforms.cpp (hunk 7 above)
Comment (zou3519) [reply aten/src/ATen/VmapTransforms.cpp:162]: Oh, that's a really good point. When I originally wrote this code I thought it was more convenient to do it this way. But now that I think about it more, it's easier to write this by iterating over `dim`.

**Comment by zou3519** [comment]
> One thing that I feel like would be quite useful for testing and understandability is a non-optimized reference implementation of the transforms, written recursively. 

I'll keep that in mind and add a task for this to the queue. I agree that it would be valuable to have a short reference implementation but I'm not sure what I can come up with would be shorter than the current implementation (which is 100 lines with comments). I spent some time actually trying to write said reference implementation but the verbosity of handling vectors in c++ quickly took the line count to more than I would have liked.

**Comment by ezyang** [comment]
Maybe we should just write the reference implementation in Haskell
